### ISSUE TITLE
Extremely slow data processing of tensorflow dataset using when using tf.while_loop inside a tf.function or normal function

### ISSUE THREAD

**Issue by prikmm** [tensorflow/tensorflow#49373]
**System information:**
Tried the code on both - Kaggle and Colab (Both only CPU)

**Code:**
I am trying to process a dataset containing ~20,000 images, their feature vectors and their masks(image segmentation). Due to the size of data, I can't perform in-memory processing.

I load the data using this function:

```
@tf.function
def get_mask_cum_featvecs_dataset(files_paths):
  paths_ds = tf.data.Dataset.list_files(files_paths)
  ds = paths_ds.interleave(lambda tfrec_path: tf.data.TFRecordDataset(tfrec_path).map(
                                  read_tfrec, num_parallel_calls=AUTO),
                               num_parallel_calls=AUTO)
  return ds
```
The preprocessing includes getting the euclidean distances between all the images present in the dataset and an input image (fed by the user), and then dividing all the euclidean distances by the max euclidean distance out of all euclidean distances present. For this I use the below given functions:
```
@tf.function
def euclidean_dist(comp_img_features, img_featvec):
    euclidean_dist = tf.sqrt(tf.reduce_sum(tf.square(img_featvec-comp_img_features), 0))
    return euclidean_dist

@tf.function
def preprocess(img_featvec, img_mask, mask_cum_featvecs_paths):
   
    tensor_img_featvec = img_featvec
    tensor_img_mask = img_mask
    tensor_mask_cum_featvecs_paths = mask_cum_featvecs_paths
    
    # the below given statements are used when this function
    # is not wrapped by tf.function decorator:
    #tensor_img_featvec = tf.constant(img_featvec)
    #tensor_img_mask = tf.constant(img_mask)
    #tensor_mask_cum_featvecs_paths = tf.constant(mask_cum_featvecs_paths)
    
    compare_img_ds = get_mask_cum_featvecs_dataset(tensor_mask_cum_featvecs_paths)
    
    get_img_euclidean_dists = partial(euclidean_dist, img_featvec=tensor_img_featvec)
    
    print("Calculating Euclidean Distances......")
    features_n_eucl_ds = compare_img_ds.map(lambda image_name, image, featvec, mask: (image_name, image, featvec, mask,
                                                                           get_img_euclidean_dists(featvec)),
                                 num_parallel_calls=AUTO)


    print("Finding Maximum Euclidean Distance........")
    num_of_loops = 2
    eucl_val_ds = features_n_eucl_ds.map(lambda image_name, image, featvec, mask, eucl_val: eucl_val,
                                         num_parallel_calls=AUTO)
   
    def dataset_reduce_max(ds, i):
        ds = ds.batch(5000)
        ds = ds.map(lambda eucl_vals: tf.math.reduce_max(eucl_vals),
                    num_parallel_calls=AUTO)
        i += 1
        return ds, i
    
    def loop_cond(ds, i):
        return tf.less(i, num_of_loops)
    
    eucl_val_ds, _ = tf.while_loop(loop_cond,
                                   dataset_reduce_max,
                                   [eucl_val_ds, tf.constant(0)])
    
    max_eucl_dist = tf.constant(0, dtype=tf.float32)

    #---------(I)---------#
    for item in eucl_val_ds.take(1):
        max_eucl_dist = item

    ratio_eucl_ds = features_n_eucl_ds.map(
                        lambda image_name, image, featvec, mask, eucl_val: (image_name,
                                                                            image,
                                                                            featvec,                                                                                             
                                                                            mask,                                                                                            
                                                                            eucl_val/max_eucl_dist),
                                           num_parallel_calls=AUTO)


    return ratio_eucl_ds #----------(II)-----------#

```
There is no in-built function for finding the maximum element in a particular datastream in tensorflow dataset (here, datastreams = image_names, image, featvecs, masks, eucl_dists) To find the maximum the euclidean distance, I created a separate tf.dataset containing euclidean distances and I used dataset reduction strategy. I created batches of size 5000, and return the maximum element present in them. Using this strategy, I can effectively get the maximum euclidean distance in 2 iterations.

**The problem:**
When I use tf.function decorator then the entire function will execute very fast and then get stuck at "return ratio_eucl_ds"(I). If I don't use the tf.function decorator then the function gets stuck at for-loop(II). It takes ~12 mins to execute the code. When I randomly put any value in "max_eucl_dist", then the preprocess function executes seamlessly. Hence, I guess the problem is with the way I am trying to extract the value for "max_eucl_dist".

Colab Link for reproducing the issue:
https://colab.research.google.com/drive/1VFEDdfImcchMYNgA68Bi989EGiHiQ2rC?usp=sharing
The notebook also contains model building code (required to get the feature vector and mask of the input image) and also two data processing functions, a tf.function decorated function and other not decorated.

**Current Behaviour:**
Not using tf.function decorator - 
```
CPU times: user 9min 35s, sys: 16min 2s, total: 25min 37s
Wall time: 13min 15s
```
Using tf.function decorator - 
```
CPU times: user 6min 28s, sys: 7min 42s, total: 14min 11s
Wall time: 9min 26s
```

But, when I randomly or manually initialize the "max_eucl_dist", then the processing the works seamlessly.
Not using tf.function decorator - 
```
CPU times: user 1.33 s, sys: 1.83 ms, total: 1.34 s
Wall time: 6.36 s
```
Using tf.function decorator -
```
CPU times: user 1.24 s, sys: 22 ms, total: 1.27 s
Wall time: 6.16 s
```

**Comment by Saduf2019** [comment]
@prikmm 
In the colab gist shared can you please use Runtime>Change Runtime>gpu and let us know if it helps.

**Comment by prikmm** [comment]
@Saduf2019 
Tried running with GPU, but encountered issue #34112, solved the error using the method provided in issue #34519. I had to remove the tf.function qualifier from both "_get_mask_cum_featvecs_dataset_" and "_preprocess_" functions:
```
def get_mask_cum_featvecs_dataset(files_paths):
  .
  .
  .
  return ds

def preprocess(img_featvec, img_mask, mask_cum_featvecs_paths):
  .
  .
  .
  return ds
```

But, like the issues #34112 and #34519 state, the dataset is a Variant type, which is placed on CPU. I guess, since the dataset is present on CPU and is not allowed to be copied to GPU, even if the GPU is connected it is not being used. All the processing is being done on CPU.
Btw, since the processing only needed 2 iterations, I tried by removing the loop and hardcoding the lines. But, still there was no improvement.

**Comment by luozhouyang** [comment]
I am confused that why you use `tf.function` when you build dataset pipelines? I think `tf.function` is not designed for this situation.

**Comment by prikmm** [comment]
Hey @luozhouyang 
At first, I didn't, but then for each image input, I had to wait too long to get the results, It was extremely frustating:grimacing:, and to get even a sight improvement if possible , I though that since, I am using `tf.while_loop` I should try to use the `tf.function` qualifier. But, it turns out I was wrong. Irrespective of the methods, the processing takes too much of time. I would like to know the why this is happening and if there is also a possible work around?

**Comment by prikmm** [comment]
Hey @rmothukuru,
Can I get an update on this issue? any work around of sort.

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
